evm/vm_state.py

```diff
@@ -0,0 +1,325 @@
+    def gas_limit(self):
+        return self.block_header.gas_limit
+
+    #
+    # chaindb
+    #
+    # @property
+    # def chaindb(self):
+    #     return self._chaindb
+
+    def set_chaindb(self, db):
+        self._chaindb = db
+
+    #
+    # state_db
+    #
+    @contextmanager
+    def state_db(self, read_only=False):
+        state = self._chaindb.get_state_db(self.block_header.state_root, read_only)
+        yield state
+
+        if read_only:
+            # This acts as a secondary check that no mutation took place for
+            # read_only databases.
+            assert state.root_hash == self.block_header.state_root
+        elif self.block_header.state_root != state.root_hash:
+            self.block_header.state_root = state.root_hash
+
+        self.access_logs.reads.update(state.db.access_logs.reads)
+        self.access_logs.writes.update(state.db.access_logs.writes)
+
+        # remove the reference to the underlying `db` object to ensure that no
+        # further modifications can occur using the `State` object after
+        # leaving the context.
+        state.db = None
+        state._trie = None
+
+    #
+    # Snapshot and Revert
+    #
+    def snapshot(self):
+        """
+        Perform a full snapshot of the current state.
+
+        Snapshots are a combination of the state_root at the time of the
+        snapshot and the checkpoint_id returned from the journaled DB.
+        """
+        return (self.block_header.state_root, self._chaindb.snapshot())
+
+    def revert(self, snapshot):
+        """
+        Revert the VM to the state at the snapshot
+        """
+        state_root, checkpoint_id = snapshot
+
+        with self.state_db() as state_db:
+            # first revert the database state root.
+            state_db.root_hash = state_root
+            # now roll the underlying database back
+
+        self._chaindb.revert(checkpoint_id)
+
+    def commit(self, snapshot):
+        """
+        Commits the journal to the point where the snapshot was taken.  This
+        will destroy any journal checkpoints *after* the snapshot checkpoint.
+        """
+        _, checkpoint_id = snapshot
+        self._chaindb.commit(checkpoint_id)
+
+    #
+    # Access self.prev_headers (Read-only)
+    #
+    def get_ancestor_hash(self, block_number):
+        """
+        Return the hash for the ancestor with the given block number.
+        """
+        ancestor_depth = self.block_header.block_number - block_number
+        if ancestor_depth > MAX_PREV_HEADER_DEPTH or ancestor_depth < 1:
+            return b''
+        header = self.get_block_header_by_hash(self.block_header.parent_hash)
+        while header.block_number != block_number:
+            header = self.get_block_header_by_hash(header.parent_hash)
+        return header.hash
+
+    def get_block_header_by_hash(self, block_hash):
+        """
+        Returns the block header by hash.
+        """
+        for value in self.prev_headers:
```

Comment:
It looks like we could store `self.prev_headers` as a mapping from `hash => header` so that we can do O(1) lookups rather than repeatedly enumerating this list to find the necessary header.

Reply:
I found that I missed that if we store `self.prev_headers` as a list is good for access `block.blockhash(uint blockNumber)` in contract code, we could make `VMState.get_ancestor_hash(block_number)` get block_hash via block_number directly since the `prev_headers` should be chained already:
```python
    def get_ancestor_hash(self, block_number):
        """
        Return the hash of the ancestor with the given block number.
        """
        ancestor_depth = self.block_header.block_number - block_number - 1
        if (ancestor_depth >= MAX_PREV_HEADER_DEPTH or
                ancestor_depth < 0 or
                ancestor_depth >= len(self.prev_headers)):            
            return b''
        header = self.prev_headers[ancestor_depth]
        return header.hash
```
Then we get O(1) for BLOCKHASH opcode.

And since `VMState` has its block properties, we can make `VMState.get_parent_header(block_header)` into `@property parent_header()` for 
 `VMState.validate_block(block)`:
```python
    @property
    def parent_header(self):
        return self.prev_headers[0]
```

But, unfortunately, `VMState.validate_uncle(uncle)` still requires calling `VMState.get_block_header_by_hash(uncle.parent_hash)`:
https://github.com/hwwhww/py-evm/blob/81b4b1ae1c854c97540b63ebd41f196c83a35dad/evm/vm/forks/frontier/vm_state.py#L305

Look on the bright side, `MAX_UNCLE_DEPTH` is 6.

With these changes, I'd vote for accessing via `block_number` directly for sharding clients. But for non-sharding clients, they may call `validate_uncle` more than call BLOCKHASH(many-ancestors-ago).

p.s. EIP 96 would make `BLOCKHASH` very different: https://github.com/ethereum/EIPs/pull/210